GoogleCloudPlatform/magic-modules
# Fix inconsistencies in how IAM policies are stored in state

# Description

Partially fixes https://github.com/hashicorp/terraform-provider-google/issues/8701

This PR changes how the `google_iam_policy` data source stores the user-entered data in its  computed `policy_data` attribute. It is addressing the first issue I describe in [this comment on the original issue](https://github.com/hashicorp/terraform-provider-google/issues/8701#issuecomment-1227543271).

I've updated how the `google_iam_policy` data source processes its `binding` blocks into a JSON string which it stores as the `policy_data` attribute. I've made this processing resemble what happens in the API, so there should not be a mismatch between the JSON of the policy in the data source's state and the resource's state.

# Manual tests

Using [the config I describe in this comment](https://github.com/hashicorp/terraform-provider-google/issues/8701#issuecomment-1227543271) I apply the first plan to create everything and then make this small edit 

```diff
-    description = "This one has a description"
+    description = "This one has a description and I edited it"
```

When I generate a plan with the latest provider version I see lots of unnecessary diffs as reported in the original issue:

![Screenshot 2022-08-25 at 17 10 03](https://user-images.githubusercontent.com/15078782/186743415-8afdf5c6-eec2-4343-af85-48f21be3535a.png)

When I generate a plan using this PR's changes the diff is much smaller:

![Screenshot 2022-08-25 at 17 20 06](https://user-images.githubusercontent.com/15078782/186743086-b1558b54-23e9-4825-98a4-6a1fc422a7ad.png)

**👆 The extra diffs for setting an empty description are another issue I'm thinking of handling in a separate PR to this one**
- Here: https://github.com/GoogleCloudPlatform/magic-modules/pull/6473

# Checklist

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] ~~Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).~~ N/A
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
    - I ran the `TestAccArtifactRegistryRepositoryIamPolicyGenerated` acceptance test that uses the `google_iam_policy` data source and it passed ok
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.


**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
iam: fixed diffs between `policy_data` from `google_iam_policy` data source and policy data in API responses
```
